pkg/typesystem/typesystem.go

```diff
@@ -667,167 +667,212 @@ func (t *TypeSystem) relationInvolvesExclusion(objectType, relation string, visi
-// then true is returned along with a nil error.
-func hasEntrypoints(
+type RelationDetails struct {
+
+	// hasEntrypoints is true when a particular object+relation relates to at least one assignable object type.
+	// A relation such as `define document#viewer: [document#viewer]` implies that `hasEntrypoints=false`.
+	// A relation such as `define document#viewer: [user, document#viewer]` implies that `hasEntrypoints=true`.
+	hasEntrypoints bool
+
+	// hasLoop is true when a particular object+relation introduces a loop in the model.
+	// hasEntrypoints=true implies that hasLoop=false.
+	// If hasEntrypoints=false, nothing can be said about hasLoop.
+	// Loops are introduced by computed usersets.
+	hasLoop bool
+}
+
+// GetRelationDetails returns information about a particular type and relation.
+func (t *TypeSystem) GetRelationDetails(
+	typeName string,
+	relationName string,
+) (*RelationDetails, error) {
+	inputRelation, _ := t.GetRelation(typeName, relationName)
+	// we can ignore the error because the getRelationDetails will check for it
+	rewrite := inputRelation.GetRewrite()
+	return getRelationDetails(t.GetAllRelations(), typeName, relationName, rewrite, map[string]map[string]*RelationDetails{})
+}
+
+func getRelationDetails(
 	typedefs map[string]map[string]*openfgav1.Relation,
 	typeName, relationName string,
 	rewrite *openfgav1.Userset,
-	visitedRelations map[string]map[string]struct{},
-) (bool, bool, error) {
+	visitedRelations map[string]map[string]*RelationDetails,
+) (*RelationDetails, error) {
 	v := maps.Clone(visitedRelations)
 
-	if val, ok := v[typeName]; ok {
-		val[relationName] = struct{}{}
-	} else {
-		v[typeName] = map[string]struct{}{
-			relationName: {},
-		}
-	}
-
 	relation, ok := typedefs[typeName][relationName]
 	if !ok {
-		return false, false, fmt.Errorf("undefined type definition for '%s#%s'", typeName, relationName)
+		return nil, fmt.Errorf("undefined type definition for '%s#%s'", typeName, relationName)
 	}
 
+	updateVisited(v, typeName, relationName)
+
 	switch rw := rewrite.Userset.(type) {
 	case *openfgav1.Userset_This:
+		// At least one type must have an entrypoint.
 		for _, assignableType := range relation.GetTypeInfo().GetDirectlyRelatedUserTypes() {
-			if assignableType.GetRelationOrWildcard() == nil || assignableType.GetWildcard() != nil {
-				return true, false, nil
+			assignableTypeName := assignableType.GetType()
+
+			if _, ok := typedefs[assignableTypeName]; !ok {
+				return nil, fmt.Errorf("undefined type '%s'", assignableTypeName)
+			} else {
+				if assignableType.GetRelationOrWildcard() == nil || assignableType.GetWildcard() != nil {
+					v[typeName][relationName].hasEntrypoints = true
+					break
+				}
 			}
 
-			assignableTypeName := assignableType.GetType()
 			assignableRelationName := assignableType.GetRelation()
 
 			assignableRelation, ok := typedefs[assignableTypeName][assignableRelationName]
```

Comment:
note (non-blocking): this kind of code adds a lot of noise to the logic herein and it really should be validated ahead of time. If we were to hit this code block the model should have been invalid in the first place, because this would imply that we had a type restriction that referenced an undefined type and/or relation.

I suggest we add a `validated` flag to the `TypeSystem` structure to indicate if a model has already been validated. The assumption is that it should have already been validated by the time we are seeing it in this code, but if it hasn't then before doing any code herein we can do

```go
func (t *TypeSystem) getRelationDetails(...) (*RelationDetails, error) {
    if !t.validated {
        t.Validate() // lazy model validation
    }

    // assume the model is now valid from here below
}
```

I suggest we use a similar lazy build/validate pattern like the wonderful @jpadilla introduced for the Conditional Relationship Tuple [EvaluableCondition](https://github.com/openfga/openfga/blob/61b87fb6cd234261308f82fda253c062d2264906/internal/condition/condition.go#L54)